### ISSUE TITLE
Find characters in range: caret may be positioned between CR & LF

### ISSUE THREAD

**Issue by Yaron10** [notepad-plus-plus/notepad-plus-plus#13531]
**STR:**
Paste the following text into a new document.
```
A
B
```
Place the caret after `A` (Pos 2).
Search -> Find characters in range.
Select "ASCII Characters".
Find.

**Result:**
The caret disappears. - It's positioned between CR & LF. 

***

Notepad++ v8.5.2   (64-bit)
Build time : Apr 11 2023 - 13:57:27
Path : ---notepad.exe
Command Line : 
Admin mode : ON
Local Conf mode : ON
Cloud Config : OFF
OS Name : Windows 10 Pro (64-bit)
OS Version : 20H2
OS Build : 19042.985
Current ANSI codepage : 1255
Plugins : 
    ComparePlus (1.1)
    HTMLTag (1.3.6)
    MenuIcons (2.0.2)
    PythonScript (2)
    _CustomizeToolbar (4.2)

**Comment by Yaron10** [comment]
Replace the `if (isFound)` block
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/ac5657d8261f9356310ad22a1cf59e1c4d1d7f02/PowerEditor/src/WinControls/FindCharsInRange/FindCharsInRange.cpp#L172
with
```cpp
	if (isFound)
	{
		auto sci_line = (*_ppEditView)->execute(SCI_LINEFROMPOSITION, found);
		(*_ppEditView)->execute(SCI_ENSUREVISIBLEENFORCEPOLICY, sci_line);
		auto posAfter = found = (*_ppEditView)->execute(SCI_POSITIONAFTER, found);
		found = (*_ppEditView)->execute(SCI_POSITIONBEFORE, found);
		(*_ppEditView)->execute(SCI_SETSEL, (direction == dirDown) ? found : posAfter, (direction == dirDown) ? posAfter : found);
	}
```
Fixes https://github.com/notepad-plus-plus/notepad-plus-plus/issues/13532 as well.

**Comment by alankilborn** [comment]
`(direction == dirDown) ? found : posAfter, (direction == dirDown) ? posAfter : found`
...
is quite an expression! :-)

**Comment by Yaron10** [comment]
Yes, it is. :)

https://github.com/notepad-plus-plus/notepad-plus-plus/blob/ac5657d8261f9356310ad22a1cf59e1c4d1d7f02/PowerEditor/src/WinControls/FindCharsInRange/FindCharsInRange.cpp#L178

Could you test/improve it in your free time?

Thank you.

**Comment by rdipardo** [comment]
> Could you test/improve it in your free time?

Like this, perhaps:

~~~cpp
  auto anchor = (direction == dirDown) ? found : posAfter;
  auto caret = (direction == dirDown) ? posAfter : found;
  (*_ppEditView)->execute(SCI_SETSEL, anchor, caret);
~~~

Since the root cause is EOL splitting, this would also work, while being a little more "self-documenting":

~~~cpp
auto sci_line = (*_ppEditView)->execute(SCI_LINEFROMPOSITION, found);
auto eol_pos = (*_ppEditView)->execute(SCI_GETLINEENDPOSITION, sci_line);
auto eol_mode = (*_ppEditView)->execute(SCI_GETEOLMODE);
auto posAfter = (found == (size_t)eol_pos && (int)eol_mode == SC_EOL_CRLF) ? found + 2: found + 1;
auto anchor = (direction == dirDown) ? found : posAfter;
auto caret = (direction == dirDown) ? posAfter : found;
(*_ppEditView)->execute(SCI_ENSUREVISIBLE, sci_line);
(*_ppEditView)->execute(SCI_GOTOPOS, found);
(*_ppEditView)->execute(SCI_SETSEL, anchor, caret);
~~~

**Comment by alankilborn** [comment]
STR:
Paste the following text into a new document (with Windows line-endings).
```
A
B
```
Place the caret after A (Pos 2).
Ctrl+f.
Put `\r` in _Find what_.
Set _Search mode_ to NOT be _Normal_.
Press _Find Next_.

Result:
The caret disappears. - It's positioned between CR & LF.

------------------------------

Similar...  Probably needs separate issue?

**Comment by alankilborn** [comment]
@rdipardo 

> this would also work, while being a little more "self-documenting"

Are you inclined to turn that into a PR, or should I?

**Comment by Yaron10** [comment]
@rdipardo,

Thank you for looking into it. 👍 

```cpp
  auto anchor = (direction == dirDown) ? found : posAfter;
  auto caret = (direction == dirDown) ? posAfter : found;
  (*_ppEditView)->execute(SCI_SETSEL, anchor, caret);
```

That indeed makes the code easier to understand.

```cpp
auto sci_line = (*_ppEditView)->execute(SCI_LINEFROMPOSITION, found);
auto eol_pos = (*_ppEditView)->execute(SCI_GETLINEENDPOSITION, sci_line);
auto eol_mode = (*_ppEditView)->execute(SCI_GETEOLMODE);
auto posAfter = (found == (size_t)eol_pos && (int)eol_mode == SC_EOL_CRLF) ? found + 2: found + 1;
auto anchor = (direction == dirDown) ? found : posAfter;
auto caret = (direction == dirDown) ? posAfter : found;
(*_ppEditView)->execute(SCI_ENSUREVISIBLE, sci_line);
(*_ppEditView)->execute(SCI_GOTOPOS, found);
(*_ppEditView)->execute(SCI_SETSEL, anchor, caret);
```

That's also very nice.
But...
1. If you place the caret at the beginning line 2 and search *upward*, the caret disappears.
That's why I've added `auto posAfter = found = ` and `found = (*_ppEditView)->execute(SCI_POSITIONBEFORE, found);`.
2. It doesn't fix https://github.com/notepad-plus-plus/notepad-plus-plus/issues/13532.

@alankilborn,

> Put \r in Find what.
...

Testing that was "requested" and smart. 👍 
Could you open an issue?

**Comment by rdipardo** [comment]
@Yaron10,
> 1. If you place the caret at the beginning of line 2 and search _upward_, the caret disappears.
>    That's why I've added `auto posAfter = found = ` and `found = (*_ppEditView)->execute(SCI_POSITIONBEFORE, found);`.
> 2. It doesn't fix [Find characters in range: multi-byte character is not selected #13532](https://github.com/notepad-plus-plus/notepad-plus-plus/issues/13532).
>

You're right! I must have misread the "Expected Behavior" description in the original ticket. . .

**Comment by Yaron10** [comment]
@rdipardo,

> I must have misread the "Expected Behavior" description in the original ticket. . .

I assumed that was the case.
I know you're highly intelligent. :)

Thanks again.
